lib/worker/exec.go

```diff
@@ -0,0 +1,129 @@
+package worker
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os/exec"
+	"strings"
+	"time"
+)
+
+// Exec represents something to be executed in the Worker
+// TODO: I might just remove this interface in the next
+// PR. I introduced it thinking it could be useful for
+// the future cgroup implementation but now am not so
+// sure.
+type Exec interface {
+	// Command is the command for the exec.
+	Command() string
+	// Start starts the exec.
+	Start() error
+	// Stop stops the exec.
+	Stop() error
+	// PID returns the pid of the process once it has been started.
+	// Otherwise, and invalid pid 0 is returned.
+	PID() int
+	// Status returns the ProcessStatus struct corresponding
+	// to the started exec.
+	Status() (*ProcessStatus, error)
+}
+
+// process is a process executed or to be executed by the worker
+// TODO: Find a better name for this as the word "process" implies
+// that this it is already running.
+type process struct {
+	command    string
+	cmd        *exec.Cmd
+	startedBy  string
+	startedAt  time.Time
+	finishedAt time.Time // TODO: implement finished at
+}
+
+func newProcess(command string, cmd *exec.Cmd, startedBy string) *process {
+	return &process{
+		command:   command,
+		cmd:       cmd,
+		startedBy: startedBy,
+	}
+}
+
+// ProcessStatus represents a snapshot of a process status
+type ProcessStatus struct {
+	PID        int
+	StartedBy  string
+	State      string
+	StartedAt  time.Time
+	FinishedAt time.Time // TODO: implement finished at
+}
+
+func (p *process) Start() error {
+	p.startedAt = time.Now()
+	err := p.cmd.Start()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *process) Stop() error {
+	if p.cmd.Process == nil {
+		return fmt.Errorf("process not started")
+	}
+	return p.cmd.Process.Kill()
+}
+
+func (p *process) Command() string {
+	return p.command
+}
+
+func (p *process) PID() int {
+	if p.cmd.Process == nil {
+		return 0
+	}
+	return p.cmd.Process.Pid
+}
+
+func (p *process) Status() (*ProcessStatus, error) {
+	if p.cmd.Process == nil {
+		return nil, fmt.Errorf("process not started")
+	}
+	pid := p.PID()
+	state, err := retrieveProcessState(pid)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ProcessStatus{
+		PID:        pid,
+		StartedBy:  p.startedBy,
+		State:      state,
+		StartedAt:  p.startedAt,
+		FinishedAt: p.finishedAt,
+	}, nil
+}
+
+// retrieveProcessState retrieves the Linux process state i.e.
+// one of R, D, S, T or Z.
+// TODO: this could belong to its own internal os package but for now I think
+// it is fine here.
+func retrieveProcessState(pid int) (string, error) {
```

Comment:
Just realized that we should probably return a default from here if the file is missing. Otherwise, we can't query for processes that already ended a while ago.

Reply:
I'm not sure we necessarily need to the underlying linux process state - we only need the state of the process that will be modeled by the worker service.

Perhaps you can maintain an internal record of the process state in the `Process` struct and use that. It would also get rid of a bunch of text parsing, at the cost of making your `Process` type a bit more complex.

Reply:
Am not sure I understand the suggestion. Do you mean, instead of using the Linux process state just maintain our own version of process state like for example "SCHEDULED" (when new process is created), "RUNNING" (when started is called), "FINISHED" (when stop is called or the process terminates)?

Reply:
If I understood the suggestion correctly. I think that would remove some observability on what the process is doing. Determining if a process is sleeping for example, doesn't seem possible with just https://pkg.go.dev/os#Process.